category_theory/natural_transformation.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2017 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tim Baumann, Stephen Morgan, Scott Morrison
+
+Defines natural transformations between functors.
+
+Introduces notations
+  `F ⟹ G` for the type of natural transformations between functors `F` and `G`,
+  `τ X` (a coercion) for the components of natural transformations,
+  `σ ⊟ τ` for vertical compositions, and
+  `σ ◫ τ` for horizontal compositions.
+-/
+
+import .functor
+
+namespace category_theory
+
+universes u₁ v₁ u₂ v₂ u₃ v₃
+
+variables {C : Type u₁} [𝒞 : category.{u₁ v₁} C] {D : Type u₂} [𝒟 : category.{u₂ v₂} D]
+include 𝒞 𝒟
+
+structure natural_transformation (F G : C ↝ D) : Type (max u₁ v₂) :=
+(components : Π X : C, (F X) ⟶ (G X))
+(naturality : ∀ {X Y : C} (f : X ⟶ Y), (F.map f) ≫ (components Y) = (components X) ≫ (G.map f) . obviously)
```

Comment:
Okay, `nat_trans` coming up. I really don't understand the desire to abbreviate. Once upon a time memory was expensive, but it's not anymore, and code is so much more readable if you use unambiguous english words, particularly the explicit mathematical terms everyone is used to.

Reply:
I've renamed `components` to `app`, though it makes me feel sick to obfuscate my code. :-)

Reply:
When I'm staring at a 30 line goal, long names make everything harder to look at. Lean already has significant term bloat, so I try to control what I can. Symbols are especially good at making things concise and readable, but they generally aren't an option when writing proofs. The names of major concepts have to be especially short because they appear as components in theorem names, and when you use a "name the symbols" approach to theorem naming those names can get long and repetitive, so short components pay for themselves.

I notice that this definition does not have a doc comment. I would much prefer you add all explanatory information to a doc comment rather than giving everything a long name.

Reply:
Okay! doc comments coming up.